File: frontend/src/types/strategy.ts
export type StrategyType = 'momentum' | 'mean_reversion' | 'trend_following' | 'factor_based' | 'custom';
export type StrategyStatus = 'draft' | 'active' | 'archived';
export type BacktestStatus = 'pending' | 'running' | 'completed' | 'failed';

export const STRATEGY_TYPE_LABELS: Record<StrategyType, string> = {
  momentum: '动量策略',
  mean_reversion: '均值回归',
  trend_following: '趋势跟踪',
  factor_based: '因子策略',
  custom: '自定义',
};

export const STRATEGY_STATUS_LABELS: Record<StrategyStatus, string> = {
  draft: '草稿',
  active: '运行中',
  archived: '已归档',
};

export const BACKTEST_STATUS_LABELS: Record<BacktestStatus, string> = {
  pending: '等待中',
  running: '运行中',
  completed: '已完成',
  failed: '失败',
};

export interface Strategy {
  id: number;
  name: string;
  code: string;
  strategy_type: StrategyType;
  status: StrategyStatus;
  description?: string;
  logic?: string;
  parameters?: Record<string, unknown>;
  is_builtin: boolean;
  created_by?: number;
  created_at: string;
  updated_at: string;
}

export interface StrategyCreate {
  name: string;
  code: string;
  strategy_type: StrategyType;
  description?: string;
  logic?: string;
  parameters?: Record<string, unknown>;
}

export interface StrategyUpdate {
  name?: string;
  code?: string;
  strategy_type?: StrategyType;
  status?: StrategyStatus;
  description?: string;
  logic?: string;
  parameters?: Record<string, unknown>;
}

export interface StrategyListParams {
  page?: number;
  size?: number;
  strategy_type?: StrategyType;
  status?: StrategyStatus;
  keyword?: string;
}

export interface StrategyListResponse {
  items: Strategy[];
  total: number;
  page: number;
  size: number;
}

export interface Backtest {
  id: number;
  strategy_id: number;
  name?: string;
  status: BacktestStatus;
  start_date: string;
  end_date: string;
  initial_capital: number;
  commission_rate: number;
  slippage: number;
  benchmark?: string;
  stock_pool?: string[];
  parameters?: Record<string, unknown>;
  error_message?: string;
  created_at: string;
  completed_at?: string;
}

export interface BacktestCreate {
  strategy_id: number;
  name?: string;
  start_date: string;
  end_date: string;
  initial_capital?: number;
  commission_rate?: number;
  slippage?: number;
  benchmark?: string;
  stock_pool?: string[];
  parameters?: Record<string, unknown>;
}

export interface BacktestResult {
  id: number;
  backtest_id: number;
  total_return?: number;
  annual_return?: number;
  benchmark_return?: number;
  alpha?: number;
  beta?: number;
  sharpe_ratio?: number;
  sortino_ratio?: number;
  max_drawdown?: number;
  volatility?: number;
  win_rate?: number;
  profit_loss_ratio?: number;
  total_trades?: number;
  avg_holding_days?: number;
  created_at: string;
}

export interface EquityCurvePoint {
  date: string;
  equity: number;
  benchmark?: number;
}

export interface BacktestDetail {
  backtest: Backtest;
  result?: BacktestResult;
  equity_curve?: EquityCurvePoint[];
}

export interface Order {
  id: number;
  backtest_id: number;
  stock_code: string;
  side: 'buy' | 'sell';
  quantity: number;
  price: number;
  filled_price?: number;
  filled_quantity?: number;
  commission?: number;
  order_time: string;
  filled_time?: string;
  signal_reason?: string;
}

File: backend/app/schemas/strategy.py
from datetime import date, datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class StrategyType(str, Enum):
    MOMENTUM = "momentum"
    MEAN_REVERSION = "mean_reversion"
    TREND_FOLLOWING = "trend_following"
    FACTOR_BASED = "factor_based"
    CUSTOM = "custom"


class StrategyStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    ARCHIVED = "archived"


class BacktestStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class OrderSide(str, Enum):
    BUY = "buy"
    SELL = "sell"


class StrategyBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=50)
    strategy_type: StrategyType = StrategyType.CUSTOM
    description: Optional[str] = None
    logic: Optional[str] = None
    parameters: Optional[Dict[str, Any]] = None


class StrategyCreate(StrategyBase):
    pass


class StrategyUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    code: Optional[str] = Field(None, min_length=1, max_length=50)
    strategy_type: Optional[StrategyType] = None
    status: Optional[StrategyStatus] = None
    description: Optional[str] = None
    logic: Optional[str] = None
    parameters: Optional[Dict[str, Any]] = None


class StrategyResponse(StrategyBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: StrategyStatus
    is_builtin: bool
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class StrategyListResponse(BaseModel):
    items: List[StrategyResponse]
    total: int
    page: int
    size: int


class BacktestCreate(BaseModel):
    strategy_id: int
    name: Optional[str] = None
    start_date: date
    end_date: date
    initial_capital: float = Field(default=1000000.0, gt=0)
    commission_rate: float = Field(default=0.0003, ge=0, le=0.01)
    slippage: float = Field(default=0.001, ge=0, le=0.05)
    benchmark: Optional[str] = "000300"
    stock_pool: Optional[List[str]] = None
    parameters: Optional[Dict[str, Any]] = None


class BacktestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    strategy_id: int
    name: Optional[str] = None
    status: BacktestStatus
    start_date: datetime
    end_date: datetime
    initial_capital: float
    commission_rate: float
    slippage: float
    benchmark: Optional[str] = None
    stock_pool: Optional[List[str]] = None
    parameters: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None


class BacktestListResponse(BaseModel):
    items: List[BacktestResponse]
    total: int


class BacktestResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    backtest_id: int
    total_return: Optional[float] = None
    annual_return: Optional[float] = None
    benchmark_return: Optional[float] = None
    alpha: Optional[float] = None
    beta: Optional[float] = None
    sharpe_ratio: Optional[float] = None
    sortino_ratio: Optional[float] = None
    max_drawdown: Optional[float] = None
    volatility: Optional[float] = None
    win_rate: Optional[float] = None
    profit_loss_ratio: Optional[float] = None
    total_trades: Optional[int] = None
    avg_holding_days: Optional[float] = None
    created_at: datetime


class EquityCurvePoint(BaseModel):
    date: str
    equity: float
    benchmark: Optional[float] = None


class BacktestDetailResponse(BaseModel):
    backtest: BacktestResponse
    result: Optional[BacktestResultResponse] = None
    equity_curve: Optional[List[EquityCurvePoint]] = None


class OrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    backtest_id: int
    stock_code: str
    side: OrderSide
    quantity: int
    price: float
    filled_price: Optional[float] = None
    filled_quantity: Optional[int] = None
    commission: Optional[float] = None
    order_time: datetime
    filled_time: Optional[datetime] = None
    signal_reason: Optional[str] = None


class OrderListResponse(BaseModel):
    items: List[OrderResponse]
    total: int


class PositionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    backtest_id: int
    stock_code: str
    quantity: int
    avg_cost: float
    market_value: Optional[float] = None
    unrealized_pnl: Optional[float] = None
    realized_pnl: Optional[float] = None


class PositionListResponse(BaseModel):
    items: List[PositionResponse]
    total: int


class StrategyTypeStats(BaseModel):
    strategy_type: StrategyType
    count: int
